src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,382 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Experimental.Collections
+{
+    /// <summary>
+    /// DictionarySlim<TKey, TValue> is similar to Dictionary<TKey, TValue> but optimized in three ways:
+    /// 1) It allows access to the value by ref.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer(assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        const int DefaultPrimeSize = 3;
+        // 1-based index into _entries; 0 means empty
+        private int[] _buckets;
+        private Entry[] _entries;
+        // 0-based index into _entries of head of free chain: -1 means empty
+        private int _freeList = -1;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            // 0-based index of next entry in chain: -1 means empty
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = new int[DefaultPrimeSize];
+            _entries = new Entry[DefaultPrimeSize];
+        }
+
+        public DictionarySlim(int capacity)
+        {
+            capacity = HashHelpers.GetPrime(capacity);
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        // Drop sign bit to ensure non negative index
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private uint GetHashCode(TKey key) => (uint)key.GetHashCode();//key.GetHashCode() & 0x7FFFFFFF;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(int bucketIndex) => _buckets[bucketIndex] - 1;
+
+        public int Count { get; private set; }
+
+        public int Capacity { get => _entries.Length; }
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
```

Comment:
Is it better not to introduce a local variable for `entries[entryIndex]`?

Reply:
no, its slower.